pkg/resources/management.cattle.io/v3/nodedriver/validator.go

```diff
@@ -0,0 +1,163 @@
+package nodedriver
+
+import (
+	"fmt"
+
+	"github.com/rancher/lasso/pkg/dynamic"
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/webhook/pkg/admission"
+	controllersv3 "github.com/rancher/webhook/pkg/generated/controllers/management.cattle.io/v3"
+	objectsv3 "github.com/rancher/webhook/pkg/generated/objects/management.cattle.io/v3"
+	admissionv1 "k8s.io/api/admission/v1"
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+var (
+	gvr = schema.GroupVersionResource{
+		Group:    "management.cattle.io",
+		Version:  "v3",
+		Resource: "nodedrivers",
+	}
+
+	driverInUse = &admissionv1.AdmissionResponse{
+		Result: &metav1.Status{
+			Status:  metav1.StatusFailure,
+			Message: "This driver is in use by existing nodes and cannot be disabled",
+		},
+		Allowed: false,
+	}
+)
+
+// Validator ValidatingWebhook for management clusters.
+type Validator struct {
+	admitter admitter
+}
+
+type admitter struct {
+	api k8sAPI
+}
+
+// interface that holds the methods necessary to hit the k8s api and fetch
+// whether rke1/2 resources are deleted
+type k8sAPI interface {
+	rke1ResourcesDeleted(driver *v3.NodeDriver) (bool, error)
+	rke2ResourcesDeleted(driver *v3.NodeDriver) (bool, error)
+}
+
+type k8sAPIImpl struct {
+	nodeCache controllersv3.NodeCache
+	dynamic   *dynamic.Controller
+}
```

Comment:
Generally, we will put our required caches and clients directly on the object and then mock out the cache and the clients in our tests. `controllersv3.NodeCache` is already, so we would just need one for `*dynamic.Controller`. This may seem like just another way to implement what you already had, but when we mock the caches and clients we get better test coverage since the current implementation leaves `rke1ResourcesDeleted` and `rke2ResourcesDeleted` untested.
```suggestion
type admitter struct {
	nodeCache controllersv3.NodeCache
	dynamic   dynamicLister
}

type dynamicLister struct {
    List(gvk schema.GroupVersionKind, namespace string, selector labels.Selector) ([]runtime.Object, error) 
}
```